### PR TITLE
refactor!: make `sid`, `id_info_id`, and `Node_ID` abstract types

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR allows `sid`, `id_info_id`, and `Node_ID` to be represented by abstract types which are different to the compiler.

## Why:
Currently, `sid`, `id_info_id`, and `Node_ID` are represented by ints.

This is fine, it's an easy and classic way to generate fresh identifiers. But it leads to some measure of unsafety, because those ints can be tampered with, and mistakes can be written in our codebase around them. By making them abstract, we guarantee that we cannot confuse them for other `int`s, or otherwise use them in a way that is not precisely described by the `MkTemp` interface.

We also duplicate some measure of code whenever we allocate our refs and re-write the counter functions. I'm sure that this is not the last we need of temporary counters, so this also helps us in the future, by standardizing our approach.

## How:
`Common` now contains a functor, `MkId`. Because OCaml has generative functors, functors which generate abstract types _always_ generate _different_ types. Thus, the module resulting from `MkId ()` and `MkId ()` have different types to the compiler, which are abstract types that are represented in the back-end by ints.

These modules provide some useful functions and have their own private counters.

## Test plan:
`make test`

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: semgrep-core/src/pfff/commons/Common.ml [hunk 1]
```diff
@@ -1639,3 +1639,38 @@ let files_of_dir_or_files_no_vcs_nofilter xs =
...
+(*****************************************************************************)
+
+(* OCaml has generative functors.
+   This means that abstract types minted from different applications of the same
+   functor are _always_ different, even if they are constructed in the exact
+   same way.
+   So different instances of the module `MkTemp` will have distinct types.
+   This helps us not conflate types when we need a new kind of temp, and avoid
+   boilerplate code with refs and counters.
+*)
```
Comment (mjambon) [inline semgrep-core/src/pfff/commons/Common.ml:1654]: What's a temp?

File: semgrep-core/src/pfff/commons/Common.mli [hunk 2]
```diff
@@ -474,3 +474,12 @@ type 'a smap = 'a SMap.t
...
 val jsoo : bool ref
+
+module MkTemp () : sig
+  type t [@@deriving show, eq, ord, hash]
+
+  val mk : unit -> t
+  val to_int : t -> int
+  val unsafe_default : t
+  val unsafe_reset_counter : unit -> unit
+end
```
Comment (mjambon) [inline semgrep-core/src/pfff/commons/Common.mli:496]: You should document the interface (try to do better than your elders...).

**Review by mjambon — APPROVED** [review]

File: semgrep-core/src/pfff/commons/Common.ml (hunk 1 above)
Comment (brandonspark) [reply semgrep-core/src/pfff/commons/Common.ml:1654]: renamed

**Comment by aryx** [comment]
Nice, I like it.

**Pull request merged**